### ISSUE TITLE
Create `State<T>` type

### ISSUE THREAD

**Issue by matthunz** [concoct-rs/concoct#13]
An efficient way of tracking state changes may be through [`Rc::make_mut`](https://doc.rust-lang.org/nightly/std/rc/struct.Rc.html#method.make_mut). Here I propose a wrapper type that impls `PartialEq` and a way of getting the state mutably, by creating a new `Rc`. This allows state propagation to be a simple pointer-equality check

Here is a rough draft, but an issue is allowing state to be used in views and modifiers that use &str, such as `struct Value<T: AsRef<str>>`. Other state mechanisms exist without the need of `Rc` but they would requiring cloning the state when a `'static` lifetime is required.
```rust
#[derive(Clone, Default)]
pub struct State<T> {
    value: Rc<Rc<RefCell<T>>>,
}

impl<T> State<T> {
    pub fn get(&self) -> Ref<T> {
        self.value.borrow()
    }

    pub fn update(&mut self) -> RefMut<T> {
        Rc::make_mut(&mut self.value).borrow_mut()
    }
}

impl<T> PartialEq for State<T> {
    fn eq(&self, other: &Self) -> bool {
        Rc::ptr_eq(&self.value, &other.value)
    }
}

// This doesn't work yet:
impl<T: AsRef<U>, U> AsRef<U> for State<T> {
    fn as_ref(&self) -> &U {
        // TODO lifetime error
        self.value.borrow().as_ref()
    }
}
```

**Comment by matthunz** [comment]
Reopening as using `Rc` solves this but at the cost of cloning the data. Without a `RefCell` or something similar large data structures will have to clone everytime to be displayed

**Comment by bitfl0wer** [comment]
> Reopening as using `Rc` solves this but at the cost of cloning the data. Without a `RefCell` or something similar large data structures will have to clone everytime to be displayed

This does not sound right. Cloning an `Rc` does not copy/clone the underlying Data, but rather just increases the internal Strong Reference Count by one while returning essentially just a pointer to the same memory location.

TL;DR: I'd love to help with this, but more information is needed :)

**Comment by matthunz** [comment]
Hey thanks for the interest!

What I think might require a clone is mutating the Rc. My plan (inspired by raph leviens blog) was to use use `Rc::make_mut` to update the state, but that will clone the data if the rc is stored in more than one place.

Still though I could see this being useful in some cases so I'd love some help figuring out how!

**Comment by bitfl0wer** [comment]
Ah, this cleared things up for me! I might think about this in the coming days and propose something if I do :)

**Comment by matthunz** [comment]
Awesome (:

**Comment by matthunz** [comment]
Closed in favor of the `use_state` hook
